### PR TITLE
Add max crypto data extension text

### PR DESCRIPTION
**Pull request by siyengar**

### CODE REVIEW

File: draft-ietf-quic-tls.md [hunk 1]
```diff
@@ -1025,6 +1025,35 @@ However, any tampering with the parameters will be detected
...
 
 
+## QUIC Max Crypto Data Extension {#max_crypto_data}
+
+When QUIC provides TLS messages via the CRYPTO_HS frame, one TLS message may be
+fragmented across different packets.  Thus TLS might have to buffer data.  TLS
+implementations may choose to limit the data they buffer before the handshake is
+completed and close the connection on receiving too much data.
+
+To communicate the maximum amount of data that TLS will allow to be sent in the
```
Comment (chris-wood) [inline draft-ietf-quic-tls.md]: remove "the' in "in the CRYPTO_HS"

File: draft-ietf-quic-tls.md [hunk 2]
```diff
@@ -1025,6 +1025,35 @@ However, any tampering with the parameters will be detected
...
+CRYPTO_HS frames, TLS may use the `max_crypto_data` extension.
```
Comment (chris-wood) [inline draft-ietf-quic-tls.md]: "... extension, defined as follows:"

File: draft-ietf-quic-tls.md [hunk 3]
```diff
@@ -1025,6 +1025,35 @@ However, any tampering with the parameters will be detected
...
+
+~~~
+   enum {
+      max_crypto_data(TBD), (65535)
+   } ExtensionType;
+
+   struct {
+	  uint32 max_crypto_data;
+   } MaxCryptoData;
```
Comment (chris-wood) [inline draft-ietf-quic-tls.md]: Maybe rename to MaxHandshakeData?

File: draft-ietf-quic-tls.md [hunk 4]
```diff
@@ -1025,6 +1025,35 @@ However, any tampering with the parameters will be detected
...
+   struct {
+	   uint32 max_crypto_data;
+   } MaxCryptoData;
+~~~
+
+The `extension_data` field of the extension contains the MaxCryptoData
+structure.
+
+Implementations SHOULD send this extension.  Receivers do not need to process
+this extension.  The purpose of this extnesion is to provide a facility to debug
```
Comment (chris-wood) [inline draft-ietf-quic-tls.md]: Typo "extnesion"

File: draft-ietf-quic-tls.md [hunk 5]
```diff
@@ -1025,6 +1025,35 @@ However, any tampering with the parameters will be detected
...
+issues during the handshake and also allow future extensibility of the protocol
```
Comment (chris-wood) [inline draft-ietf-quic-tls.md]: I would remove the note about debugging. It's primarily about controlling the amount of buffering and for future extensibility, right?

File: draft-ietf-quic-transport.md [hunk 6]
```diff
@@ -3935,6 +3935,14 @@ errors is not mandatory, but only because requiring that an endpoint generate
 these errors also means that the endpoint needs to maintain the final offset
 state for closed streams, which could mean a significant state commitment.
 
+## Flow control of CRYPTO_HS data {#flow-control-crypto}
+
+Data sent in CRYPTO_HS frames is not flow controlled in the same way as STREAM
+frames.  QUIC relies on the cryptographic protocol implementation to avoid
+excessive buffering of data, see {{QUIC-TLS}}.  The cryptographic implementation
```
Comment (chris-wood) [inline draft-ietf-quic-transport.md]: s/of data, see/of data; See

Comment (chris-wood) [inline draft-ietf-quic-transport.md]: Remove "cryptographic" from "cryptographic implementation"

File: draft-ietf-quic-tls.md [hunk 7]
```diff
@@ -1025,6 +1025,35 @@ However, any tampering with the parameters will be detected
...
+
+   struct {
+	   uint32 max_crypto_data;
+   } MaxCryptoData;
+~~~
+
+The `extension_data` field of the extension contains the MaxCryptoData
+structure.
+
+Implementations SHOULD send this extension.  Receivers do not need to process
```
Comment (chris-wood) [inline draft-ietf-quic-tls.md]: Wait -- why don't receivers need to process this? Shouldn't they limit the amount of data they send based on this limit?

**Comment by ekr** [comment]
This looks good. Two other points:

I think @kazuho wanted a separate draft. Personally, I am good either way.

You will need an IANA registration for the new extension code point.

File: draft-ietf-quic-tls.md (hunk 7 above)
Comment (siyengar) [reply draft-ietf-quic-tls.md]: no not really, if they cant fit their entire data into this limit then, they're screwed anyway.

File: draft-ietf-quic-tls.md (hunk 4 above)
Comment (siyengar) [reply draft-ietf-quic-tls.md]: ??

Comment (siyengar) [reply draft-ietf-quic-tls.md]: ah

File: draft-ietf-quic-tls.md (hunk 5 above)
Comment (siyengar) [reply draft-ietf-quic-tls.md]: its mostly about debugging

File: draft-ietf-quic-tls.md (hunk 7 above)
Comment (chris-wood) [reply draft-ietf-quic-tls.md]: So if receivers don't honor the limit, what's the point?

Comment (chris-wood) [reply draft-ietf-quic-tls.md]: Chatted in Slack -- merge it!

**Pull request merged**